### PR TITLE
2020 update

### PR DESCRIPTION
**Pull request by tom91136**
See description in each commit.
Highlights:
 * Fixed bad local memory use so hipSYCL now works fine.
 * No more compiler warnings on all three compilers with the exception of ComputeCpp on kernel names
 * Added arg support for picking/listing SYCL devices
 * Fixed broken CMake flags: Release didn't actually get -O3 and all that

### CODE REVIEW

File: src/field_summary.cpp [hunk 1]
```diff
@@ -115,15 +115,15 @@ void field_summary(global_variables &globals, parallel_ &parallel) {
 					           result.buffer);
 				},
 				[](const ctx &ctx, id<1> lidx) { ctx.local[lidx] = {}; },
-				[ymax, ymin, xmax, xmin](ctx ctx, id<1> lidx, id<1> idx) {
+				[ymin, xmax, xmin](const ctx &ctx, id<1> lidx, id<1> idx) {
```
Comment (tomdeakin) [inline src/field_summary.cpp:118]: Is this strictly `const` when `ctx.actual.*` is written?

File: src/PdV.cpp [hunk 2]
```diff
@@ -83,10 +83,6 @@ void PdV_kernel(
 
 			double volume_change_s = volume[idx] / (volume[idx] + total_flux);
 
-			double min_cell_volume =
```
Comment (tomdeakin) [inline src/PdV.cpp]: I'm worried these are actually needed. We can't delete the computation as it needs to match exactly all the versions.

File: src/PdV.cpp [hunk 3]
```diff
@@ -147,18 +139,11 @@ void PdV_kernel(
 //  @details Invokes the user specified kernel for the PdV update.
 void PdV(global_variables &globals, bool predict) {
 
-	double kernel_time;
+	double kernel_time = 0;
 	if (globals.profiler_on) kernel_time = timer();
 
 	globals.error_condition = 0;
 
-	int prdct;
```
Comment (tomdeakin) [inline src/PdV.cpp:155]: This looks important

File: src/advec_cell.cpp [hunk 4]
```diff
@@ -129,7 +126,6 @@ void advec_cell_kernel(
 						sigma3 = (1.0 + sigmat) * (vertexdx[j] / vertexdx[dif]);
 						sigma4 = 2.0 - sigmat;
 
-						sigma = sigmat;
```
Comment (tomdeakin) [inline src/advec_cell.cpp]: We should keep this for similar reasons to above.

File: src/advec_cell.cpp [hunk 5]
```diff
@@ -262,7 +254,6 @@ void advec_cell_kernel(
 						sigma3 = (1.0 + sigmat) * (vertexdy[k] / vertexdy[dif]);
 						sigma4 = 2.0 - sigmat;
 
-						sigma = sigmat;
```
Comment (tomdeakin) [inline src/advec_cell.cpp]: Keep.

**Review by tomdeakin — CHANGES_REQUESTED** [review]
Generally great improvements, thanks!
We need to not delete bits of the compute, even if the results are unused.

File: src/field_summary.cpp (hunk 1 above)
Comment (tom91136) [reply src/field_summary.cpp:118]: It's a const reference so it only prevents reassignment of the variable itself, but that's better than nothing; it matches every other use of lambdas in the a reduction call

File: src/PdV.cpp (hunk 3 above)
Comment (tom91136) [reply src/PdV.cpp:155]: `prdct` is dead code I think, there were no further references to that and it's not inside a kernel or kernel capture

File: src/PdV.cpp (hunk 2 above)
Comment (tom91136) [reply src/PdV.cpp]: I'll keep this but I'm pretty sure it gets compiled away as `min_cell_volume` was never used after declaration.

File: src/advec_cell.cpp (hunk 4 above)
Comment (tom91136) [reply src/advec_cell.cpp]: Will keep, although it was created, assigned, and never used again and it doesn't escape the kernel it any way.

**Pull request merged**